posts/2024-02-16-Clippy-deprecating-feature-cargo-clippy.md

```diff
@@ -0,0 +1,87 @@
+---
+layout: post
+title: "Clippy: Deprecating `feature = \"cargo-clippy\"`"
+author: The Clippy Team
+release: true
+---
+
+Since Clippy [`v0.0.97`] and before it was shipped with `rustup`, Clippy
+implicitly added a `feature = "cargo-clippy"`[^1] when linting your code with
+`cargo clippy`.
+
+[^1]: It's likely, that you didn't even know that Clippy implicitly sets this
+    feature. This is intentional, as we stopped advertising and documenting this
+    a long time ago.
+
+Back in the day (2016) this was necessary to allow, warn or deny lints with
+attributes:
+
+```rust
+#[cfg_attr(feature = "cargo-clippy", allow(clippy_lint_name))]
+```
+
+Doing this hasn't been necessary for a long time. Today, Clippy users will set
+lint levels with tool lint attributes using the `clippy::` prefix:
+
+```rust
+#[allow(clippy::lint_name)]
+```
+
+The implicit `feature = "cargo-clippy"` has only been kept for backwards
+compatibility, but will now be deprecated.
+
+## Alternative
+
+As there is a rare [use case] for conditional compilation depending on Clippy,
+we will provide an alternative. So in the future you will be able to use:
+
+```rust
+#[cfg(clippy)]
+```
+
+## Transitioning
+
+Should you have instances of `feature = "cargo-clippy"` in your code base, you
+will see a warning from the new Clippy lint
+[`clippy::deprecated_clippy_cfg_attr`][pr-12292]. This lint can automatically fix
+your code. So if you should see this lint triggering, just run:
+
+```
+cargo clippy --fix -- -Aclippy::all -Wclippy::deprecated_clippy_cfg_attr
+```
+
+This will fix all instances in your code.
+
+In addition, check your `.cargo/config` file for:
+
+```toml
+[target.'cfg(feature = "cargo-clippy")']
+rustflags = ["-Aclippy::..."]
+```
+
+This file you have to update yourself, by either changing it to `cfg(clippy)` or
+taking this opportunity to transition to [setting lint levels in
+`Cargo.toml`][cargo-lints] directly.
+
+## Motivation for Deprecation
+
+There's a plan to stabilize [checking conditional compilation at compile
+time][rfc-3013], aka `cargo check -Zcheck-cfg`. If we were to keep `feature =
```

Comment:
We may want to mention the [Call for Testing](https://github.com/rust-lang/rfcs/pull/3013#issuecomment-1936648479).

Reply:
Good idea, added in https://github.com/rust-lang/blog.rust-lang.org/pull/1253/commits/f56b2c68d73cef8856123c1cebfebdffed77cbf0